02_runtime/10-metadata.adoc

```diff
@@ -0,0 +1,727 @@
+ * stem:[t_r] is the type Id (<<defn-rtm-type-id>>) of the runtime.
+
+.Polkadot Runtime Metadata
+[kaitai#metadata,kaitai-dependencies="metadata_type,metadata_pallet,metadata_extrinsic"]
+++++
+seq:
+  - id: magic
+    contents: meta
+  - id: metadata_version
+    type: u1
+
+  - id: num_types
+    type: scale::compact_int
+  - id: types
+    type: metadata_type
+    repeat: expr
+    repeat-expr: num_types.value
+
+  - id: num_pallets
+    type: scale::compact_int
+  - id: pallets
+    type: metadata_pallet
+    repeat: expr
+    repeat-expr: num_pallets.value
+
+  - id: extrinsic_type
+    type: scale::compact_int
+  - id: extrinsic_version
+    type: u1
+  - id: num_extrinsics
+    type: scale::compact_int
+  - id: extrinsics
+    type: metadata_extrinsic
+    repeat: expr
+    repeat-expr: num_extrinsics.value
+
+  - id: runtime_type
+    type: scale::compact_int
+++++
+
+.Runtime Registry Type Entry
+[#defn-rtm-registry-entry]
+====
+
+A registry entry contains information about a type in its portable form for
+serialization. The entry is a datastructure of the following format:
+
+[stem]
+++++
+r_i = (i_t,p,T,D,c) \
+T = (t_0, ...,t_n) \
+t_i = (n, y)
+++++
+
+where::
+* stem:[i_t] is a compact integer indicating the identifier of the type.
+* stem:[p] is the path of the type, optional and based on source file location. Encoded a sequence (<<defn-scale-list>>) of strings.
+* stem:[T] is a sequence (<<defn-scale-list>>) of generic parameters (empty for non-generic types).
+** stem:[n] is the name string of the generic type parameter
+** stem:[y] is a _Option_ type containing a type Id (<<defn-rtm-type-id>>).
+* stem:[D] is the type definition (<<defn-rtm-type-definition>>).
+* stem:[c] is the documentation as sequence (<<defn-scale-list>>) of strings.
+
+.Runtime Metadata Type
+[kaitai#metadata_type,kaitai-dependencies="metadata_type_definition"]
+++++
+seq:
+  - id: id
+    type: scale::compact_int
+
+  - id: path
+    type: scale::string_list
+
+  - id: num_params
+    type: scale::compact_int
+  - id: params
+    repeat: expr
+    repeat-expr: num_params.value
+    type: param
+
+  - id: definition
+    type: metadata_type_definition
+
+  - id: docs
+    type: scale::string_list
+types:
+  param:
+    seq:
+      - id: name
+        type: scale::string
+      - id: type
+        type: scale::maybe_compact_int
+++++
+====
+
+.Runtime Type Id
+[#defn-rtm-type-id]
+====
+The runtime type Id is a compact integer representing the index to the
+type in the type registry (<<defn-rtm-registry-entry>>).
+
+====
+
+.Type Variant
+[#defn-rtm-type-definition]
+====
+The type definition stem:[D] indicates all the possible types of encodable
+values a type can have.
+
+[stem]
+++++
+D = {
+	(0,->,C,"composite type (e.g. structure or tuple)"),
+	(1,->,V,"variant type"),
+	(2,->,s_v,"sequence type varying length"),
+	(3,->,S,"sequence with fixed length"),
```

Comment:
Maybe nice to be consistent with the MetadataTypeDefinition diagram and add 'Array' in brackets next to D=3 -> Sequence with fixed length. In general make notations in Definition of D consistent with the diagram

Reply:
I guess so, but this should be updated in the Kaitai structure. We use the word "sequence" almost exclusively.

Reply:
Yes, that is my mistake. The kaitai struct definition was partial written of polkadot-js json data, so I might have messed up some of the names. Feel free to change the definition.